### DISCUSSION TITLE
Kerberos Authentication

### DISCUSSION THREAD

**Discussion by Silverdust-ZH** [ontop/ontop#815] [Q&A]
Hello everyone
I need to connect to an Oracle Database that is protected by a Kerberos Authentication Schema. I am using an ojdbc8-21.3.0.0.jar driver but I could not find any property related to Kerberos in https://ontop-vkg.org/guide/advanced/configuration.html to tell ontop to use Kerberos as Authentication Schema. Did I miss something? Also, searching for "Kerberos" within the source files yields no result.

Kind regards
Samuel

**Comment by bcogrel** [comment]
Hi @Amadeus82, Ontop has nothing specific  to Kerberos but it gives full control to the user over the JDBC URL as well as over the properties passed to the JDBC connection (see `jdbc.property.<KEY>` in https://ontop-vkg.org/guide/advanced/configuration).

Shouldn't it be enough? I don't have experience with Kerberos for Oracle.

If you get something working, please feel free to share it, so we can enrich the page on Oracle: https://ontop-vkg.org/guide/databases/oracle .

**Comment by ahjulstad** [comment]
Reviving an old thread just to describe the approach that I have working:

Short story, when Oracle drivers and Java are compatible and correctly configured, with kerberos credentials available in the credentials cache, it just works without username and password, both in the Protege plugin and from the endpoint. 

The steps are a bit complicated, though, but this works for me. 

In a debian bookworm devcontainer  (bookworm because of Oracle OCI driver compatibility), add Java.

```
{
	"name": "Debian",
	"image": "mcr.microsoft.com/devcontainers/base:bookworm",
	"features": {
		"ghcr.io/devcontainers/features/java:1": {
			"version": "21",
			"jdkDistro": "tem",
			"gradleVersion": "latest",
			"mavenVersion": "latest",
			"antVersion": "latest",
			"groovyVersion": "latest",
			"installMaven": "true"
		}
	},
	"postCreateCommand": "./install_oracle.sh && ./install_ontop.sh",
	"containerEnv": {
		"LD_LIBRARY_PATH": "/opt/instantclient_21_20",
		"TNS_ADMIN": "${containerWorkspaceFolder}/tns_admin"
	}
}
```

Add Oracle drivers:  (`install_oracle.sh`)   I found that 21.20 is a recent version where binaries and JDBC JAR are happy. 

```
wget -nc -P /tmp https://download.oracle.com/otn_software/linux/instantclient/2120000/instantclient-basic-linux.x64-21.20.0.0.0dbru.zip
sudo unzip -o /tmp/instantclient-basic-linux.x64-21.20.0.0.0dbru.zip -d /opt
echo "/opt/instantclient_21_20" | sudo tee /etc/ld.so.conf.d/oracle-instantclient.conf

sudo ldconfig

sudo apt update
sudo DEBIAN_FRONTEND=noninteractive apt install -y -q krb5-user sssd libaio1
```

Add Ontop and JDBC:  (`install_ontop.sh`)

```

wget -nc -P /tmp https://github.com/ontop/ontop/releases/download/ontop-5.5.0/ontop-cli-5.5.0.zip
sudo unzip -o /tmp/ontop-cli-5.5.0.zip -d /opt/ontop 

sudo ln -s /opt/ontop/ontop /usr/local/bin/ontop

sudo wget -nc -P /opt/ontop/jdbc https://repo1.maven.org/maven2/com/oracle/database/jdbc/ojdbc11/21.20.0.0/ojdbc11-21.20.0.0.jar

sudo ln -s /opt/instantclient_21_20/libocijdbc21.so /usr/lib
```

You need a sqlnet.ora file, telling the Oracle client to use Kerberos. The file is located in a path given by the TNS_ADMIN environment variable (I just set it in a custom run script)    (the krb5.conf file might not be necessary)

```
SQLNET.AUTHENTICATION_SERVICES = (KERBEROS5)
SQLNET.KERBEROS5_CONF          = /etc/krb5.conf
SQLNET.KERBEROS5_CONF_MIT      = TRUE
```

Then, you need to populate the Kerberos credentials cache with an initial login:

```
kinit <USERNAME>@<REALM>
```

The run script:

```
export TNS_ADMIN=./tns_admin 
ontop endpoint -m xx_ontology.obda -t xx_ontology.ttl -p xx_ontology.properties  --cors-allowed-origins=*
```
